File: app/src/main/java/com/elementary/tasks/core/additional/QuickSmsActivity.java
package com.elementary.tasks.core.additional;

import android.app.PendingIntent;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.telephony.SmsManager;
import android.view.WindowManager;

import com.elementary.tasks.R;
import com.elementary.tasks.core.ThemedActivity;
import com.elementary.tasks.core.data.models.SmsTemplate;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.utils.Contacts;
import com.elementary.tasks.core.utils.LogUtil;
import com.elementary.tasks.core.utils.Permissions;
import com.elementary.tasks.core.utils.SuperUtil;
import com.elementary.tasks.core.view_models.sms_templates.SmsTemplatesViewModel;
import com.elementary.tasks.databinding.ActivityQuickSmsLayoutBinding;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class QuickSmsActivity extends ThemedActivity {

    private static final int REQ_SMS = 425;

    private SelectableTemplatesAdapter mAdapter;

    private String number;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initData();
        ActivityQuickSmsLayoutBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_quick_sms);

        binding.messagesList.setLayoutManager(new LinearLayoutManager(this));
        mAdapter = new SelectableTemplatesAdapter(this);
        binding.messagesList.setAdapter(mAdapter);

        binding.buttonSend.setOnClickListener(v -> startSending());
        String name = Contacts.getNameFromNumber(number, this);
        binding.contactInfo.setText(SuperUtil.appendString(name, "\n", number));

        initViewModel();
    }

    private void initViewModel() {
        SmsTemplatesViewModel viewModel = ViewModelProviders.of(this).get(SmsTemplatesViewModel.class);
        viewModel.smsTemplates.observe(this, smsTemplates -> {
            if (smsTemplates != null) {
                updateList(smsTemplates);
            }
        });
    }

    private void updateList(List<SmsTemplate> smsTemplates) {
        mAdapter.setData(smsTemplates);
        if (mAdapter.getItemCount() > 0) {
            mAdapter.selectItem(0);
        }
    }

    private void initData() {
        number = getIntent().getStringExtra(Constants.SELECTED_CONTACT_NUMBER);
    }

    private void startSending() {
        if (!Permissions.checkPermission(this, Permissions.SEND_SMS)) {
            Permissions.requestPermission(this, REQ_SMS, Permissions.SEND_SMS);
            return;
        }
        int position = mAdapter.getSelectedPosition();
        SmsTemplate item = mAdapter.getItem(position);
        if (item != null) {
            LogUtil.d("TAG", "startSending: " + item.getTitle());
            sendSMS(number, item.getTitle());
        }
        removeFlags();
    }

    public void removeFlags() {
        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
                | WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                | WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON
                | WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD);
        finish();
    }

    private void sendSMS(String number, String message) {
        String SENT = "SMS_SENT";
        String DELIVERED = "SMS_DELIVERED";
        PendingIntent sentPI = PendingIntent.getBroadcast(this, 0, new Intent(SENT), 0);
        PendingIntent deliveredPI = PendingIntent.getBroadcast(this, 0, new Intent(DELIVERED), 0);
        SmsManager sms = SmsManager.getDefault();
        sms.sendTextMessage(number, null, message, sentPI, deliveredPI);
    }

    @Override
    public void onBackPressed() {
        removeFlags();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (grantResults.length == 0) return;
        switch (requestCode) {
            case REQ_SMS:
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    startSending();
                }
                break;
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/additional/QuickSmsActivity.kt
package com.elementary.tasks.core.additional

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.WindowManager
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.elementary.tasks.R
import com.elementary.tasks.core.analytics.Feature
import com.elementary.tasks.core.analytics.FeatureUsedEvent
import com.elementary.tasks.core.arch.BindingActivity
import com.elementary.tasks.core.data.models.SmsTemplate
import com.elementary.tasks.core.utils.Constants
import com.elementary.tasks.core.utils.contacts.Contacts
import com.elementary.tasks.core.utils.Permissions
import com.elementary.tasks.core.utils.TelephonyUtil
import com.elementary.tasks.core.utils.nonNullObserve
import com.elementary.tasks.core.view_models.sms_templates.SmsTemplatesViewModel
import com.elementary.tasks.databinding.ActivityQuickSmsBinding
import org.koin.androidx.viewmodel.ext.android.viewModel

class QuickSmsActivity : BindingActivity<ActivityQuickSmsBinding>() {

  private val viewModel by viewModel<SmsTemplatesViewModel>()
  private var mAdapter: SelectableTemplatesAdapter = SelectableTemplatesAdapter()
  private var number: String = ""

  override fun inflateBinding() = ActivityQuickSmsBinding.inflate(layoutInflater)

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    number = intent.getStringExtra(Constants.SELECTED_CONTACT_NUMBER) ?: ""

    binding.messagesList.layoutManager = LinearLayoutManager(this)
    binding.messagesList.adapter = mAdapter

    binding.buttonSend.setOnClickListener { startSending() }
    val name = if (Permissions.checkPermission(this, Permissions.READ_CONTACTS)) {
      Contacts.getNameFromNumber(number, this) ?: ""
    } else {
      ""
    }
    binding.contactInfo.text = "$name\n$number"

    analyticsEventSender.send(FeatureUsedEvent(Feature.QUICK_SMS))

    initViewModel()
  }

  private fun initViewModel() {
    viewModel.smsTemplates.nonNullObserve(this) { smsTemplates ->
      updateList(smsTemplates)
    }
  }

  private fun updateList(smsTemplates: List<SmsTemplate>) {
    mAdapter.setData(smsTemplates)
    if (mAdapter.itemCount > 0) {
      mAdapter.selectItem(0)
    }
  }

  private fun startSending() {
    val position = mAdapter.selectedPosition
    val item = mAdapter.getItem(position)
    if (item != null) {
      sendSMS(number, item.title)
    } else {
      sendError()
    }
  }

  private fun removeFlags() {
    window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
      or WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
      or WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON
      or WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD)
    finish()
  }

  private fun sendSMS(number: String?, message: String?) {
    if (number != null) {
      TelephonyUtil.sendSms(this, number, message)
    } else {
      sendError()
    }
  }

  private fun sendError() {
    Toast.makeText(this, R.string.error_sending, Toast.LENGTH_SHORT).show()
  }

  override fun handleBackPress(): Boolean {
    removeFlags()
    return true
  }

  companion object {

    fun openScreen(context: Context, number: String) {
      context.startActivity(Intent(context, QuickSmsActivity::class.java)
        .putExtra(Constants.SELECTED_CONTACT_NUMBER, number)
        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
    }
  }
}
